Add page with users from database

EugeneTurkin opened the pull request:
After we've created a registration page (issue https://github.com/EugeneTurkin/forum123/issues/5), the next step could be adding functionality for it. Like form validation and creating users in database. But before that it's going to be easier if we already will have an ability to check all registered users on forum. With such feature we can easily make sure that our registration works fine. So we need to add the ability to see registered users before we start to implement registration.

So in the scope of this task we need to add basic, let's say it "index page" where will be listed all users we have in database.

### Steps to do:

PostgreSQL:
- add `docker-compose.yml` file which sets up `PostgreSQL` database using docker containers

SQLAlchemy:
- add `sqlalchemy` to `requirements.txt` and setup it locally
- add `database.py` which initializes `sqlalchemy` by creating `engine`, `session` and `Base`

Alembic:
- add `alembic` to `requirements.txt` and setup it locally
- initialize and configure `alembic` for our project using `alembic init alembic` command inside project root directory
- connect `alembic` to `PostgreSQL` using `sqlalchemy.url` option inside `alembic.ini`
- connect `alembic` to `sqlalchemy`'s metadata using `target_metadata` variable inside `alembic`'s `env.py`

Models:
- add `models.py` with a simple `User` model with fields: `id`, `username`
- register this model in `alembic`'s `env.py` via importing it below the import of `Base`
- create a migration for `User` model using `alembic revision --autogenerate -m "add users table"`
- apply created migration using `alembic upgrade <revision>`
- add two users to database using  `sqlalchemy`'s model `User` in python interactive shell
- check that users have been added successfully using another query via `sqlalchemy` in python interactive shell

Route:
- update implementation of "index" page to make it show all users from database
for example it's going to be a page with something like this: 
> Welcome to forum123!
>
> We already have 2 users joined us:
>     - John
>     - Bob

Other:
- update `README.md` with information about how to run database container and apply migrations to it

birthdaysgift commented on README.md:
```diff
@@ -44,3 +44,15 @@ Linters order above is the preffered way to run and fix them one by one.
+
+### Run database container
+
+1. In terminal type in `docker-compose -f envs/dev/docker-compose.yml up --detach` command. (you need Docker and Docker Compose installed)
```
Let's split this info into two sections. Info about Docker and Docker Compose requirements can be added to the top of the `README.md` file - alongside with python version:
> This project requires Python 3.10, Docker, Docker Compose.

And command we need to run the database can be in `Run application` section, because our application cannot work without database. Let's put it after the first item:
> 1. Open terminal.
> 2. Run services using `docker-compose -f envs/dev/docker-compose.yml up --detach` command
> ...

Also, let's add info about applying migrations to database, because database needs migrations to work correctly. Like this:
> 1. Open terminal.
> 2. Run services using `docker-compose -f envs/dev/docker-compose.yml up --detach` command
> 3. Apply migrations if you haven't done it yet: `alembic ugrade head`
> ...

birthdaysgift commented on README.md:
```diff
@@ -44,3 +44,15 @@ Linters order above is the preffered way to run and fix them one by one.
+
+### Apply migrations to database
+
+1. In terminal type in `alembic upgrade <revision id>`. (Revision ID may be found in forum123/alembic/versions/<revision>)
```
I've found in the [docs](https://alembic.sqlalchemy.org/en/latest/tutorial.html#running-our-first-migration) that we can run `alembic upgrade head` if we want to apply all migrations up to the last one.
So we can change this command to `alembic upgrade head`

birthdaysgift commented on README.md:
```diff
@@ -44,3 +44,15 @@ Linters order above is the preffered way to run and fix them one by one.
+
+
+
+### Apply migrations to database
```
Let's name this section "Migrations" or "Working with migrations" and add more info to it. Let's add info about how to:
- apply migrations (`upgrade`)
- revert migrations (`downgrade`)
- autogenerate migrations

birthdaysgift commented on src/templates/index.html:
```diff
@@ -0,0 +1,11 @@
+        <li>{{ user.username }}</li>
+    {% endfor %}
+</ul>
+{% endblock %}
```
Let's add a newline at the end of the file

birthdaysgift commented on src/routes.py:
```diff
@@ -13,7 +15,8 @@
 @bp.route("/index")
 def index() -> str:
     """Use this view function to check whether Flask is installed properly."""
```
Since this route is used more than just checking Flask installation, let's update docstring. We can use an approach we applied for `registration` route, like
```python
"""Handle index page."""
```

birthdaysgift commented on src/models.py:
```diff
@@ -0,0 +1,14 @@
+    __tablename__ = "users"
+
+    id = Column(Integer, primary_key=True)  # noqa: A003
+    username = Column(String)
```
We totally forgot to add database constraints to our fields. Here username should be`unique=True`, because we will use `username` field as a unique identificator for user, and `nullable=False` because we cannot have a user without `username`.
Yep, to accomplish that you need to do some routine again:
- run database
- revert migrations
- remove migration file
- add constraints to `models.py`
- generate new migration
- apply migration to database
- create two users
- check that `/index` works

birthdaysgift commented on src/database.py:
```diff
@@ -0,0 +1,14 @@
+from sqlalchemy.orm import declarative_base, scoped_session, sessionmaker
+
+# will allow us to connect to PostgreSQL database
+engine = create_engine("postgresql://postgres:@localhost:5432/postgres")
```
Let's make connection string to be configurable.

What we want here is to have an ability to provide different connection strings to our application without rewriting code. Why do we need this? Because our application may run on different databases depending on environment, for example:
- for dev environment we need `postgresql://postgres:@localhost:5432/postgres`
- for test environment we will need `postgresql://testuser:somepassword@test-stand-url.com:5432/testdb`
- for prod environment we will need `postgresql://superuser:supersecretpassword@prod-stand-url:5432/forum123_db`

So we need an ability to provide different connection strings in different environments. 

Feel free to ask any questions if you feel confused about "how" and "what" you need to do for this fix.

birthdaysgift commented on alembic.ini:
```diff
@@ -0,0 +1,66 @@
+# version_path_separator = space
+version_path_separator = os  # Use os.pathsep. Default configuration used for new projects.
+
+sqlalchemy.url = postgresql://postgres:@localhost:5432/postgres
```
Here we need to solve same problem which is described [here](https://github.com/EugeneTurkin/forum123/pull/17/files#r1066802661). We will discuss it together when you fix the problem at the place where link points to.

EugeneTurkin committed:
6: Add page with users from database

After we've created a registration page (issue https://github.com/EugeneTurkin/forum123/issues/5), the next step could be adding functionality for it. Like form validation and creating users in database. But before that it's going to be easier if we already will have an ability to check all registered users on forum. With such feature we can easily make sure that our registration works fine. So we need to add the ability to see registered users before we start to implement registration.

So in the scope of this task we need to add basic, let's say it "index page" where will be listed all users we have in database.

PostgreSQL:
- add `docker-compose.yml` file which sets up `PostgreSQL` database using docker containers

SQLAlchemy:
- add `sqlalchemy` to `requirements.txt` and setup it locally
- add `database.py` which initializes `sqlalchemy` by creating `engine`, `session` and `Base`

Alembic:
- add `alembic` to `requirements.txt` and setup it locally
- initialize and configure `alembic` for our project using `alembic init alembic` command inside project root directory
- connect `alembic` to `PostgreSQL` using `sqlalchemy.url` option inside `alembic.ini`
- connect `alembic` to `sqlalchemy`'s metadata using `target_metadata` variable inside `alembic`'s `env.py`

Models:
- add `models.py` with a simple `User` model with fields: `id`, `username`
- register this model in `alembic`'s `env.py` via importing it below the import of `Base`
- create a migration for `User` model using `alembic revision --autogenerate -m "add users table"`
- apply created migration using `alembic upgrade <revision>`
- add two users to database using  `sqlalchemy`'s model `User` in python interactive shell
- check that users have been added successfully using another query via `sqlalchemy` in python interactive shell

Route:
- update implementation of "index" page to make it show all users from database
for example it's going to be a page with something like this:
> Welcome to forum123!
>
> We already have 2 users joined us:
>     - John
>     - Bob

Other:
- update `README.md` with information about how to run database container and apply migrations to it
